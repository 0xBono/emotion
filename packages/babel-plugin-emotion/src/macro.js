// @flow
import { transformExpressionWithStyles } from '@emotion/babel-utils'
import { buildMacroRuntimeNode, addRuntimeImports } from './babel-utils'
import { createMacro } from 'babel-plugin-macros'

export default createMacro(macro)

function macro({ references, state, babel }) {
  let t = babel.types
  Object.keys(references).forEach(referenceKey => {
    let isPure = true
    switch (referenceKey) {
      case 'injectGlobal': {
        isPure = false
      }
      // eslint-disable-next-line no-fallthrough
      case 'css':
      case 'keyframes': {
        references[referenceKey].reverse().forEach(reference => {
          const path = reference.parentPath
          const runtimeNode = buildMacroRuntimeNode(
            reference,
            state,
            referenceKey,
            t
          )
          reference.replaceWith(runtimeNode)
          if (isPure) {
            path.addComment('leading', '#__PURE__')
          }
          transformExpressionWithStyles({
            babel,
            state,
            path,
            shouldLabel: state.opts.autoLabel
          })
        })
        break
      }
      default: {
        references[referenceKey].reverse().forEach(reference => {
          reference.replaceWith(
            buildMacroRuntimeNode(reference, state, referenceKey, t)
          )
        })
      }
    }
  })
  addRuntimeImports(state, t)
}
