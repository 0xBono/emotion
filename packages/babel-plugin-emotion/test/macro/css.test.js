// @flow
import 'test-utils/legacy-env'
import React from 'react'
import renderer from 'react-test-renderer'
import { css } from 'emotion/macro'

test('float property', () => {
  const cls1 = css`
    float: left;
  `

  const tree = renderer.create(<div className={cls1} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('handles more than 10 dynamic properties', () => {
  const cls1 = css`
    text-decoration: ${'underline'};
    border-right: solid blue 54px;
    background: ${'white'};
    color: ${'black'};
    display: ${'block'};
    border-radius: ${'3px'};
    padding: ${'25px'};
    width: ${'500px'};
    z-index: ${100};
    font-size: ${'18px'};
    text-align: ${'center'};
  `

  const tree = renderer.create(<div className={cls1} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('auto px', () => {
  const cls1 = css({ display: 'flex', flex: 1, fontSize: 10 })
  const tree = renderer.create(<div className={cls1} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('random interpolation with undefined values', () => {
  const cls2 = css`
    ${undefined};
    justify-content: center;
  `
  const tree = renderer.create(<div className={cls2} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('random expression', () => {
  const cls2 = css`
    font-size: 20px;
    @media (min-width: 420px) {
      color: blue;
      ${css`
        width: 96px;
        height: 96px;
      `};
      line-height: 40px;
    }
    background: green;
  `
  const tree = renderer.create(<div className={cls2} />).toJSON()
  expect(tree).toMatchSnapshot()
})
