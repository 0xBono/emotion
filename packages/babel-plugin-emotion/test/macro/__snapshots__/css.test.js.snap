// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto px 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 10px;
}

<div
  className="emotion-0"
/>
`;

exports[`float property 1`] = `
.emotion-0 {
  float: left;
}

<div
  className="emotion-0"
/>
`;

exports[`handles more than 10 dynamic properties 1`] = `
.emotion-0 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border-right: solid blue 54px;
  background: white;
  color: black;
  display: block;
  border-radius: 3px;
  padding: 25px;
  width: 500px;
  z-index: 100;
  font-size: 18px;
  text-align: center;
}

<div
  className="emotion-0"
/>
`;

exports[`random expression 1`] = `
.emotion-0 {
  font-size: 20px;
  background: green;
}

@media (min-width:420px) {
  .emotion-0 {
    color: blue;
    width: 96px;
    height: 96px;
    line-height: 40px;
  }
}

<div
  className="emotion-0"
/>
`;

exports[`random interpolation with undefined values 1`] = `
.emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  className="emotion-0"
/>
`;
